### PR TITLE
Fix webhook timing issue while installing clustertasks and triggers

### PR DESCRIPTION
**Pull request by nikhil-thomas**
Modify directory v0.8.0 payload

Modify v0.8.0 payload directory structure

Rename config.operator.tekton.dev status codes

Signed-off-by: Nikhil Thomas <nikthoma@redhat.com>

### CODE REVIEW

File: pkg/controller/validate/validate.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package validate
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
```
Comment (hrishin) [inline pkg/controller/validate/validate.go]: any reason leaving `blankspace`?

File: pkg/controller/config/config_controller.go [hunk 2]
```diff
@@ -152,34 +165,46 @@ func (r *ReconcileConfig) Reconcile(req reconcile.Request) (reconcile.Result, er
 		return reconcile.Result{}, err
 	}
 
-	if isUpToDate(cfg) {
+	if stateCheck(cfg, op.InstalledStatus) {
 		log.Info("skipping installation, resource already up to date")
 		return reconcile.Result{}, nil
 	}
-
+	if stateCheck(cfg, op.ValidatedPipeline) {
```
Comment (hrishin) [inline pkg/controller/config/config_controller.go]: :+1: for idiomatic status checks.

Additionally wouldn't be nice flow would be organic?
- [ ] installing pipelines if not installed
- [ ] validate pipeline if installing
- [ ] install add on if pipeline is validated

File: pkg/controller/config/config_controller.go [hunk 3]
```diff
@@ -228,6 +253,101 @@ func (r *ReconcileConfig) reconcileInstall(req reconcile.Request, cfg *op.Config
...
+			Code:    op.ErrorStatus,
+			Details: err.Error(),
+			Version: flag.TektonVersion})
+		return reconcile.Result{}, err
+	}
+
+	if err := r.addons.ApplyAll(); err != nil {
+		log.Error(err, "failed to apply addons yaml manifest")
+		// ignoring failure to update
+		_ = r.updateStatus(cfg, op.ConfigCondition{
```
Comment (hrishin) [inline pkg/controller/config/config_controller.go]: Wondering if failure in stalling add-on should mark pipeline `Config` in error state?

File: pkg/apis/operator/v1alpha1/config_types.go [hunk 4]
```diff
@@ -45,15 +45,35 @@ type ConfigCondition struct {
...
+	InstallingAddons InstallStatus = "installing-addons"
+
+	// InstalledStatus indicates that all pipeline resources are installed successfully
+	InstalledStatus InstallStatus = "installed"
 
 	// ErrorStatus indicates that there was an error installing pipeline resources
 	// Check details field for additional details
 	ErrorStatus InstallStatus = "error"
+
+	// InstallingStatus indicates that the pipeline resources are being installed
```
Comment (sthaha) [inline pkg/apis/operator/v1alpha1/config_types.go]: Delete?

File: pkg/apis/operator/v1alpha1/config_types.go [hunk 5]
```diff
@@ -45,15 +45,35 @@ type ConfigCondition struct {
...
 
-	// InstallingStatus indicates that the pipeline resources are being installed
-	InstallingStatus InstallStatus = "installing"
+	// AppliedPipeline indicates that the core pipeline resources
+	// have been applied on the cluster
+	AppliedPipeline InstallStatus = "applied-core-pipeline"
+
+	// WaitingPipelineValidation indicates that core pipeline deployments are being
+	// ensured to be running
+	WaitingPipelineValidation InstallStatus = "waiting-pipeline-validation"
```
Comment (sthaha) [inline pkg/apis/operator/v1alpha1/config_types.go]: how about `validating-pipeline` ?

File: pkg/controller/config/config_controller.go [hunk 6]
```diff
@@ -3,18 +3,19 @@ package config
...
+	"path/filepath"
+	"time"
 
 	"github.com/go-logr/logr"
 	mf "github.com/jcrossley3/manifestival"
 	sec "github.com/openshift/client-go/security/clientset/versioned/typed/security/v1"
 	"github.com/operator-framework/operator-sdk/pkg/predicate"
 	"github.com/prometheus/common/log"
 	op "github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	"github.com/tektoncd/operator/pkg/controller/transform"
```
Comment (sthaha) [inline pkg/controller/config/config_controller.go]: could you please move non controllers code outside the `pkg/controller` ?

File: pkg/controller/config/config_controller.go [hunk 7]
```diff
@@ -43,22 +44,33 @@ func init() {
 // Add creates a new Config Controller and adds it to the Manager. The Manager will set fields on the Controller
 // and Start it when the Manager is Started.
 func Add(mgr manager.Manager) error {
-	m, err := mf.NewManifest(flag.ResourceDir, flag.Recursive, mgr.GetClient())
```
Comment (sthaha) [inline pkg/controller/config/config_controller.go]: 🌟✨ 🦸‍♂️

File: pkg/controller/config/config_controller.go [hunk 8]
```diff
@@ -43,22 +44,33 @@ func init() {
...
 		return err
 	}
-	return add(mgr, newReconciler(mgr, m))
+	return add(mgr, rec)
 }
 
 // newReconciler returns a new reconcile.Reconciler
-func newReconciler(mgr manager.Manager, m mf.Manifest) reconcile.Reconciler {
+func newReconciler(mgr manager.Manager) (reconcile.Reconciler, error) {
+	pplnPath := filepath.Join(flag.ResourceDir, "pipelines")
```
Comment (sthaha) [inline pkg/controller/config/config_controller.go]: nit: pipelinePath; doesn't that read better?

File: pkg/controller/config/config_controller.go [hunk 9]
```diff
@@ -43,22 +44,33 @@ func init() {
...
+	ppln, err := mf.NewManifest(pplnPath, flag.Recursive, mgr.GetClient())
```
Comment (sthaha) [inline pkg/controller/config/config_controller.go]: nit: ppln -> pipeline 🙏

File: pkg/controller/config/config_controller.go [hunk 10]
```diff
@@ -152,34 +165,46 @@ func (r *ReconcileConfig) Reconcile(req reconcile.Request) (reconcile.Result, er
 		return reconcile.Result{}, err
 	}
 
-	if isUpToDate(cfg) {
+	if stateCheck(cfg, op.InstalledStatus) {
```
Comment (sthaha) [inline pkg/controller/config/config_controller.go]: How about calling it `isStatus(cfg. op.InstalledStatus)` also how about `cfg.IsStatus(op.InstalledStatus)` ?

File: pkg/controller/config/config_controller.go (hunk 2 above)
Comment (sthaha) [reply pkg/controller/config/config_controller.go]: I was thinking of a state machine like this .. 

```
switch cfg.Status() {
case ErrorStatus: return applyPipeline()
case  UnknownStatus: return applyPipeline()
case  ApplyingPipeline: return applyPipeline()
case AppliedPipeline: return validatePipeline()
case ValidatingPipeline: return validatePipeline()
case ValidatedPipeline: return applyAddons()
case AppliedAddons: return markInstalled() 
case InstalledState:     return validateUptoDate() // checks uptodate ? including uuid matching
}
```

File: pkg/controller/config/config_controller.go (hunk 3 above)
Comment (sthaha) [reply pkg/controller/config/config_controller.go]: yes, it should. These addons are mandatory and isn't managed by a different CR; so we have to indicate that failure to apply these addons as a failure.

File: pkg/controller/validate/validate.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+		Name:      name,
+	}
+	err := c.Get(ctx, key, dp)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if dp.Status.AvailableReplicas == *dp.Spec.Replicas {
```
Comment (sthaha) [inline pkg/controller/validate/validate.go]: `*` is that needed?

Comment (sthaha) [reply pkg/controller/validate/validate.go]: also `return dp.Status.AvailableReplicatas == dp.Status.Replicas , nil` removes the need for  `if`

File: pkg/controller/config/config_controller.go [hunk 12]
```diff
@@ -398,16 +522,19 @@ func createCR(c client.Client) error {
 	return err
 }
 
-func isUpToDate(cfg *op.Config) bool {
+func stateCheck(cfg *op.Config, code op.InstallStatus) bool {
+	if !matchesUUID(cfg.Status.OperatorUUID) {
```
Comment (sthaha) [inline pkg/controller/config/config_controller.go]: this need not be checked for all states. If you follow the state machine above, we can move the stateCheck and the version check to just `Installed` state.

File: pkg/controller/config/config_controller.go (hunk 2 above)
Comment (sthaha) [reply pkg/controller/config/config_controller.go]: IMHO; this state machine should suffice ... 

```
   switch cfg.InstallStatus() {
    case op.ErrorStatus:
      return r.applyPipeline(req, cfg)
    case op.EmptyStatus:
      return r.applyPipeline(req, cfg)
    case op.AppliedPipeline:
      return r.validatePipeline(req, cfg)
    case op.ValidatedPipeline:
      return r.applyAddons(req, cfg)
    case op.InstalledStatus:
      return r.validateVersion(req, cfg)
```

**Pull request merged**